CONTRIBUTING.md

```diff
@@ -3,11 +3,70 @@ Guidelines for Contributing
-* Review Phobos additions in the [Review Queue](http://wiki.dlang.org/Review_Queue).
\ No newline at end of file
+Quick links
+-----------
+
+- Fork [on Github](https://github.com/D-Programming-Language/phobos)
+- Use our [Bugzilla bug tracker](http://d.puremagic.com/issues/)
+- Follow the [Styleguide](http://dlang.org/dstyle.html)
+- Participate in [our forum](http://forum.dlang.org/)
+- Ask questions on `#d` ([Web interface](https://kiwiirc.com/client/irc.freenode.net/d))
+- Review Phobos additions in the [Review Queue](http://wiki.dlang.org/Review_Queue)
+
+How to make a good submission?
+------------------------------
+
+The usual workflow is:
+
+1. [Fork](https://help.github.com/articles/fork-a-repo/) [Phobos](https://github.com/D-Programming-Language/phobos) on Github
+2. Create your [own branch](https://github.com/Kunena/Kunena-Forum/wiki/Create-a-new-branch-with-git-and-manage-branches)
+3. Work locally on your new feature or fix (see the tips below)
+4. Submit your [pull request](https://help.github.com/articles/creating-a-pull-request/)
+
+Below is a list of tips and checks that you should consider before opening your PR:
+
+### Code
+
+- Don't add `@nogc`, `@pure`, `@safe`, `nothrow` attributes yourself - let the compiler infer it! However add unittest check for these attributes if possible
+- Use nested structs for ranges (aka [Voldemort types](http://wiki.dlang.org/Voldemort_types))
+- Avoid to use `auto` as return type wherever possible (makes the documentation easier to read)
+- Don't be afraid to use the new features (e.g. `Foo!Range` vs. `Foo!(Range)`) - some parts of Phobos are a bit older and have some relicts due to lacking functionality of the compiler
+- Be critical when looking at Phobos code
+- Don't mix bug fixes with enhancements or refactoring
+- Bigger changes (rule of thumb: >500 LoC) have to go to `std.experimental` first.
+- [ ] Did you stick to the [style guide](http://dlang.org/dstyle.html)? (e.g. braces in their own lines (allman style) or white spaces between operators)
+- [ ] Is there an easier/more performant way to do it?
+- [ ] Does your PR address only one topic? Can it be split up in smaller PRs?
+- [ ] Is your code flexible enough to accommodate more use cases?
+- [ ] Did you use clear variable names?
+- [ ] Read through your code again - is there any part that is not understandable?
+
+### Tests
+
+- [Autotester](http://wiki.dlang.org/Git_Commit_Tester) will automatically compile the code in your PR and test it on all supported platforms. For you first PR you need approval from any reviewer - ping them in case they forget.
+- [ ] Do all tests pass? You can run the tests of a single module with `rdmd -main -unittest module.d`, of an entire package with `make -f posix.mak std/algorithm.test` and all tests with `make -f posix.mak unittest`
+- [ ] Do your tests cover all cases? (you can check code coverage in the resulting `module.lst` of `rdmd -cov -main -unittest module.d`)
```

Comment:
Might want to mention that it's useful to run tests locally as well, to prevent waste of autotester cycles by silly mistakes like typos and syntax errors, or code that obviously doesn't work even locally. Phobos unittests can be run by:
````
make -f posix.mak unittest
````
But that takes a long time, so I often just run:
````
dmd -main -unittest std/algorithm/iteration.d -of/tmp/test && ./test; echo $?
````
and it will print `0` if all goes well, otherwise there will be an error message or a non-zero error code (usually means a crash).

Of course, this is on Posix only... Windows users probably would have a different workflow.